motoko/persistent-storage/README.md

```diff
@@ -0,0 +1,134 @@
+
+This example covers:
+
+- Create new canister smart contract using Motoko
+- Add backend functions for a counter (increment, get count and reset count)
+- Deploy the canister smart contract locally
+- Test backend with Candid UI and command line using DFX 
+
+## Installation
+This example project can be cloned, installed and deployed locally, for learning and testing purposes. The instructions are based on running the example on either macOS or Linux, but when using WSL2 on Windows, the instructions will be the same.
+
+### Prerequisites
+The example project requeres the following installed:
+
+- GitHub
+- DFX
+
+GitHub can be installed from various package managers. DFX can be installed following the instructions [here](https://smartcontracts.org/docs/quickstart/local-quickstart.html#download-and-install).
+
+### Install
+Install the example dapp project:
+
+```bash
+$ git clone https://github.com/dfinity/examples
+$ cd motoko/persistent storage
+```
+
+## Documentation
+The two main parts of the example dapp is the backend and the Candid interface. This example project does not have a frontend.
+
+### Motoko backend
+The backend functions are located in the `src/persistent_storage/main.mo` Motoko file. The backend stores the counter value, and has functions to get, increment and reset the counter value. Furthermore the backend insures the counter value persists upgrades of the dapp.
+
+#### Counter variable
+The current counter value is stored as a number in the actor.
+
+```javascript
+actor {
+    stable var counter : Nat = 0;
+}
+```
+
+#### increment()
+The `increment()` function increments the counter variable.
+
+```javascript
+public func increment() : async Nat {
+    counter += 1;
+    return counter;
+};
+```
+
+The function is returning the incremented counter variable.
+
+#### get()
+The `get()` function returns the current counter value.
+
+```javascript
+public query func get() : async Nat {
+    return counter;
+};
+```
+
+#### reset()
+The `reset()` function resets the counter value to 0 and returns the value.
+
+```javascript
+public func reset() : async Nat {
+    counter := 0;
+    return counter;
+};
+```
+
+### Candid interface
+The Candid interface is automatically created, and it has a convienient UI, which provides an easy, user friendly way to test the backend. The UI is also automatically generated, and the canister ID can be found in the `canister_ids.json` file.
+
+The localhost version of the `canister_ids.json` file can be found in `.dfx/local/canister_ids.json` and the URL is:
+
+**http://<candid_canister_id>.localhost:8000/?id=<backend_canister_id>**
+
+## Deployment
+The local network is started by running this command:
+
+```bash
+$ dfx start --background
+```
+
+When the local network is up and running, run this command to deploy the canisters:
+
+```bash
+$ dfx deploy
```

Comment:
Do you think these instructions should go through the process of changing the Motoko file (so that the wasm changes) in some way, redeploying, and demonstrating that the pre-upgrade counter value survived the upgrade?

Reply:
I'll cover that more in details in another example dapp, so I think I'll save it for there.